### ISSUE TITLE
Fidelity Score in matchms

### ISSUE THREAD

**Issue by hechth** [matchms/matchms#209]
**Is your feature request related to a problem? Please describe.**
In terms of expanding matchms as discussed [here](https://github.com/matchms/matchms/issues/205), the Fidelity score would be a great candidate since it is also available as metric in [METLIN](https://metlin.scripps.edu/landing_page.php?pgcontent=new_ms_ms_spectrum_scoring_search).

**Describe the solution you'd like**
An implementation of the Fidelity score similar to `CosineGreedy` or `CosineHungarian` as a subclass of `BaseSimilarity`, implementing the pair function would be a good solution.

**Any good starting points?**
An implementation of the score is already available in [Daphnis](https://github.com/YuanyueLi/Daphnis/blob/0699f720b11b6ec0d4b82f96f7d3f9f88a617578/spectral_similarity/math_distance.py#L102) and in [philentropy](https://github.com/drostlab/philentropy/blob/960847ab7126dc5347c32a873386c42d58a545d8/src/distances.h#L1039) for python and R respectively.

**Additional context**
The Fidelity score is also used in METLIN for matching mass spectra and was chosen as possible first metric from Daphnis to be implemented in matchms.

There are varying definitions on the Fidelity score, making it either the simple or squared [Bhattacharyya coefficient](https://en.wikipedia.org/wiki/Bhattacharyya_distance#Bhattacharyya_coefficient). For the purpose of spectral similarity matching, these scores deliver the same ranking, so any form is acceptable, it just has to be clearly stated which form is used and what are the consequences (or lack thereof).

There is also a possible variation in the sense of forward and reverse matching regarding the transformation from mass spectrum to probability density function (pdf).
For a forward matching (experimentally measured spectra into library), it makes sense to use all peaks for computation of the pdf, since then only a _true match_ (meaning identical intensity distribution over all m/z) will deliver a score of 1.
For reverse matching (library into experimental spectra), it would be reasonable to only use the m/z values with non zero intensity in the library (in this case the query) spectrum to create the pdf of the experimentally measured spectrum. If all non-zero intensity peaks of the experimental (in this case reference) spectrum would be used to build the pdf, the similarity score would very likely be very small, due to the presence of other peaks in the experimental spectrum. This on the other hand doesn't serve the purpose of determining the presence of the library (query) spectrum in the experimental data, since the score is lowered **not only by mismatch, but by co-presence of other peaks in the spectrum.**

**Scientific reference**
The metric was formulated originally in [Jozsa1994](https://www.tandfonline.com/doi/abs/10.1080/09500349414552171), an additional reference describing various metrics is [Cha2007](https://www.naun.org/main/NAUN/ijmmas/mmmas-49.pdf).

**Comment by sdrogers** [comment]
Morning all,
Thanks for this - some observations from me...

1. This is just a dot product after the spectra have been a) normalised so that intensities sum to one (so as to be interpreted as a probability mass function) and b) the intensities have been subsequently square rooted. These two parts feel more like normalisation than similarity — does it make sense for them to be incorporated into a similarity?
2. This presumably needs the same alignment stage as cosine right? So, do we make up the score greedily or with e.g. Hungarian?
3. Matchms doesn’t currently have the notion of forward and reverse matching — everything is symmetric. I think adding forward / backward would be a big change and require a lot of thought (maybe refactoring to add the notion of _symmetric_ and _asymmetric_ scores?). I guess reverse matches are more of a DIA thing?

**Comment by hechth** [comment]
> This presumably needs the same alignment stage as cosine right? So, do we make up the score greedily or with e.g. Hungarian?

Could you explain this a bit more in detail, which kind of alignment? Is this referring to which peak actually get's matched to which, so which terms are eventually multiplied?

**Comment by sdrogers** [comment]
That’s right - there are two general ways for performing these similarity calculations:
1. Bin the spectra into a fixed set of bins. The spectra is now a vector, and similarity between two spectra (using the same bin positions) is simple.
2. _Align_ the spectra: for each peak in spectrum A (doesn’t matter which), find the corresponding peak(s) in spectrum B (within some tolerance). Now, because each peak in A could match to more than one in B (or vice versa) and we can only _use_ each peak once when computing similarity, we have to find some optimal set of pairs....this is often done greedily, but can also be solved via the Hungarian algorithm for maximal matching in a bipartite graph.

I also thought of another thing w.r.t the _reverse_ that is relevant to the alignment discussion. If we want to do a reverse match, we have to find the peaks in spectrum A (experimental) that are in B (library). If we’re using alignment then this can only be done after alignment. At the moment we store the peak intensity products when we find the peak pairs for alignment. But in this reverse scenario, we don’t know the intensities of spectrum A until we’ve done the alignment (because we have to normalise them to make a pmf with only those that matched something in B).

Doing all that is possible, but it would require a bit of a diversion from the current structure.

So, in summary:
1. Adding the new score looks straightforward and sensible. My only Q is in whether the pmf normalisation and the sqrt of the intensities is best done inside a scoring function, or before.
2. The reverse matching requires a fair bit of thinking so suggest this is treated separately.

**Comment by hechth** [comment]
> That’s right - there are two general ways for performing these similarity calculations:
> 1. Bin the spectra into a fixed set of bins. The spectra is now a vector, and similarity between two spectra (using the same bin positions) is simple.
> 2. Align the spectra: for each peak in spectrum A (doesn’t matter which), find the corresponding peak(s) in spectrum B (within some tolerance). Now, because each peak in A could match to more than one in B (or vice versa) and we can only use each peak once when computing similarity, we have to find some optimal set of pairs....this is often done greedily, but can also be solved via the Hungarian algorithm for maximal matching in a bipartite graph.

This sounds like something that is executed independently from the actual similarity metric. Maybe we should decouple this from the actual similarity? Since there are various algorithms for computing the match (as you already said, but there are also others like ICP etc. which can be used, or even a k-nn based multiple matching would be possible and then using the final score to choose the best - there are many possibilities), this should be an own step it seems.

I also agree that the discussion on forward/reverse should be continued in a different place and is an additional topic.

**Comment by hechth** [comment]
I had a look at the code and seems like the Hungarian alignment is what would make the most sense for now - to take the peak matches in the given tolerance which would maximize the score.

The code for this on the other hand lies in the `CosineHungarian` class and would have to be extracted from there into an own namespace/file.

**Comment by florian-huber** [comment]
> The code for this on the other hand lies in the CosineHungarian class and would have to be extracted from there into an own namespace/file.

Avoiding unnecessary code duplication would also be my preference. So if the same code is used in another score we could move it to a small helper function, e.g. add it to the already existing `spectrum_similarity_functions.py` for now.

**Comment by hechth** [comment]
@florian-huber I will create an issue for that, I think it is appropriate.

**Comment by sdrogers** [comment]
Ah, I see where the other issue came from now! I can’t speak for @florian-huber, but I suggest that the Hungarian algorithm is just factored out as a method that takes as argument the two spectra, the list of pairs and a callable that takes a pair and compute what would be its contribution to the score.

apologies if that’s what the other issue is actually suggesting!!

**Comment by florian-huber** [comment]
Saw the new issue. Looks fine to me. If I get it right, what Simon suggested is in line with #210, so factoring out the Hungarian algorithm based peak matching and using this from within `CosineHungarian`. Would you do a PR on this @hechth? I'm happy to review!

**Comment by hechth** [comment]
> apologies if that’s what the other issue is actually suggesting!!

@sdrogers Yes, sorry if that wasn't clear from how I formulated it in #210  👍
